### DISCUSSION TITLE
Creazione nuova tipologia di documento

### DISCUSSION THREAD

**Discussion by Lucaccc** [iisgiua/giuaschool#676] [Richieste di aiuto]
Buon pomeriggio.
Volevo sapere se fosse possibile aggiungere una nuova tipologia tra i documenti caricabili. In caso affermativo, potrebbe dirmi quali sono le cartelle/file/path sulle quali devo agire?
Grazie mille

**Comment by trinko** [comment]
Tutto si può fare... Ma richiede molte modifiche...

In pratica è necessario:
- definire una nuova lettera per codificare il tipo di documento (vedi **src/Entity/Documento.php**)
- implementare la gestione per la visualizzazione, l'inserimento, la modifica e la cancellazione del nuovo tipo di documento (vedi il file **src/Controller/DocumentiController.php**); potrebbe essere necessaria anche qualche modifica sulle funzioni comuni usate in **src/Util/DocumentiUtil.php**
- implementare (o modificare) il form per l'inserimento/modifica dei dati, vedi **src/Form/DocumentoType.php**
- implementare/modificare i template usati per le operazioni di visualizzazione e inserimento/modifica, vedi la cartella **templates/documenti**

**Reply by Lucaccc** [reply]
Ho fatto queste modifiche e il registro funziona, quindi presumo non ci siano errori.
Non capisco, però, come fare apparire la voce di mio interesse nel menù dei documenti..

AGGIORNAMENTO - Ho provato ad inserire una linea come quella dei programmi (con i vari nomi modificati) nel file "menu_documenti.html.twig", ma in quel caso mi da errore 500 il registro quando clicco sulla voce "documenti"..

**Comment by trinko** [comment]
Il file da modificare è quello corretto.
La riga da aggiungere ha la seguente sintassi:
```
    <li role="presentation"{{ app.request.attributes.get('_route') == 'NOME' ? ' class="active"' }}><a href="{{ path('NOME') }}" title="DESCRIZIONE">TITOLO</a></li>
```
Dove:
- *TITOLO*: è il testo che appare nel menu
- *DESCRIZIONE*: è la descrizione che appare quando ci si ferma con il mouse sopra l'opzione del menu, senza cliccarci sopra
- *NOME* : è il nome univoco della pagina, che corrisponde a quanto indicato nel tag **@route** del controller, nella proprietà **name**. Ad esempio:
```   @Route("percorso_pagina", name="nome_pagina", methods={"GET"})```

Non dimenticare di cancellare la cache, dopo la modifica.

**Comment by Lucaccc** [comment]
> Il file da modificare è quello corretto. La riga da aggiungere ha la seguente sintassi:
> 
> ```
>     <li role="presentation"{{ app.request.attributes.get('_route') == 'NOME' ? ' class="active"' }}><a href="{{ path('NOME') }}" title="DESCRIZIONE">TITOLO</a></li>
> ```
> 
> Dove:
> 
> * _TITOLO_: è il testo che appare nel menu
> * _DESCRIZIONE_: è la descrizione che appare quando ci si ferma con il mouse sopra l'opzione del menu, senza cliccarci sopra
> * _NOME_ : è il nome univoco della pagina, che corrisponde a quanto indicato nel tag **@route** del controller, nella proprietà **name**. Ad esempio:
>   `   @Route("percorso_pagina", name="nome_pagina", methods={"GET"})`
> 
> Non dimenticare di cancellare la cache, dopo la modifica.

Ho fatto come ha detto ma continua a darmi errore 500 quando clicco sulla voce "documenti" dal menù del docente, nonostante il registro funzioni normalmente..

**Comment by trinko** [comment]
Carica il log generato con il debug attivo, altrimenti non posso aiutarti.

**Reply by Lucaccc** [reply]
[2024-03-29_14-19.log](https://github.com/iisgiua/giuaschool/files/14804681/2024-03-29_14-19.log)

**Comment by trinko** [comment]
L'errore è nel file **templates/ruolo_docente/menu_documenti.html.twig**: 
```
Unable to generate a URL for the named route "NOME" as such route does not exist.
```
Nel file hai lasciato scritto **NOME** al posto del nome identificativo della pagina, cioè quello che devi indicare nel tag @route del controller, nella proprietà name (come indicato in precedenza).

**Comment by Lucaccc** [comment]
> > L'errore è nel file **templates/ruolo_docente/menu_documenti.html.twig**:
> > ```
> > Unable to generate a URL for the named route "NOME" as such route does not exist.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Nel file hai lasciato scritto **NOME** al posto del nome identificativo della pagina, cioè quello che devi indicare nel tag @route del controller, nella proprietà name (come indicato in precedenza).

Ho corretto e ora si vede correttamente la pagina, però continua a darmi errore 500 quando clicco sulla nuova voce aggiunta (le altre funzionano correttamente).
Allego log errori. 
[2024-04-02_21-53.log](https://github.com/iisgiua/giuaschool/files/14843017/2024-04-02_21-53.log)

**Comment by trinko** [comment]
```
"Unable to generate a URL for the named route "documenti_pagellino_add" as such route does not exist."
```
L'errore è nel file **templates/documenti/pagellini.html.twig**, in linea 39, dove fai riferimento alla pagina **"documenti_pagellino_add"** che però non risulta ancora definita.

Vedi se nel Controller hai indicato la pagina con lo stesso nome usato nel template (tag _@route_, proprietà _name_).

**Reply by Lucaccc** [reply]
Grazie. Ho risolto quel problema.
Ora, però, mi da lo stesso errore 500 quando clicco su conferma, dopo che ho caricato il documento.

Log: 
[2024-04-03_22-16.log](https://github.com/iisgiua/giuaschool/files/14858103/2024-04-03_22-16.log)

**Comment by trinko** [comment]
L'errore è questo:
```
Cannot rename because the target FILES/archivio/classi/5M/.pdf already exists.
```
Al momento in cui si deve spostare il file caricato dalla cartella temporanea a quella della classe non viene trovato il nome del file...

Dovrebbe dipendere dalla funzione **impostaUnAllegato()** presente nel file "src/Util/DocumentiUtil.php": qui devi aggiungere il nuovo tipo di file per assegnare anche in questo caso il nome e titolo.

**Comment by Lucaccc** [comment]
> L'errore è questo:
> 
> ```
> Cannot rename because the target FILES/archivio/classi/5M/.pdf already exists.
> ```
> 
> Al momento in cui si deve spostare il file caricato dalla cartella temporanea a quella della classe non viene trovato il nome del file...
> 
> Dovrebbe dipendere dalla funzione **impostaUnAllegato()** presente nel file "src/Util/DocumentiUtil.php": qui devi aggiungere il nuovo tipo di file per assegnare anche in questo caso il nome e titolo.

Buongiorno. 
Volevo procedere come hai detto, ma stamattina ho lo stesso problema di prima: quando clicco sulla voce aggiunta mi da di nuovo errore 500...ho provato a leggere il log ma non riesco a trovare il problema.
Lo allego.
Grazie per l'aiuto.
[2024-04-05_10-29.log](https://github.com/iisgiua/giuaschool/files/14882378/2024-04-05_10-29.log)

AGGIORNAMENTO - Ho apportato le modifiche che hai detto e la voce aggiunta continua a dare errore 500. 
Ora il log è questo:
[2024-04-05_17-26.log](https://github.com/iisgiua/giuaschool/files/14887083/2024-04-05_17-26.log)

**Comment by trinko** [comment]
Il primo log riporta l'errore:
```
Attempted to call an undefined method named "pagelliniDocente" of class "App\Util\DocumentiUtil"." at src/Controller/DocumentiController.php line 531
```
Quindi il metodo **pagelliniDocente()** non è definito ma viene richiamato in **DocumentiController.php** alla riga 531.

Il secondo log riporta l'errore:
```
Undefined method "pagellini"
```
Di nuovo, viene richiamato il metodo pagellini() ma questo non risulta definito.

**Comment by Lucaccc** [comment]
> Il primo log riporta l'errore:
> 
> ```
> Attempted to call an undefined method named "pagelliniDocente" of class "App\Util\DocumentiUtil"." at src/Controller/DocumentiController.php line 531
> ```
> 
> Quindi il metodo **pagelliniDocente()** non è definito ma viene richiamato in **DocumentiController.php** alla riga 531.
> 
> Il secondo log riporta l'errore:
> 
> ```
> Undefined method "pagellini"
> ```
> 
> Di nuovo, viene richiamato il metodo pagellini() ma questo non risulta definito.

Non capisco quale sia il motivo: in src/util/DocumentiUtil.php ho inserito la funzione relativa ai pagellini, definendo pagelliniDocente. 
Per quale motivo non funziona?
C'è altro che va modificato?
Riporto il log di stamani, magari emerge altro..
Grazie.

Log: 
[2024-04-06_11-35.log](https://github.com/iisgiua/giuaschool/files/14893114/2024-04-06_11-35.log)

**Comment by trinko** [comment]
L'errore attuale riguarda il metodo **pagellini** non **pagelliniDocente**.

Controlla dove fai una chiamata al metodo pagellini: dal log sembra che il problema riguardi l'esecuzione dell'EntityRepository.
Probabilmente è dovuto all'uso del form, controlla le modifiche fatte nel file **src/Form/DocumentoType.php**
